src/transformers/tokenization_utils.py

```diff
@@ -770,3 +772,172 @@ def _decode(
             return clean_text
         else:
             return text
+
+
+class PreTrainedSentencepieceTokenizer(PreTrainedTokenizer):
+    """
+    Base class for all slow sentencepiece tokenizers.
+
+    Inherits from :class:`~transformers.tokenization_utils.PreTrainedTokenizer`.
+
+    Args:
+        vocab_file (:obj:`str`):
+            `SentencePiece <https://github.com/google/sentencepiece>`__ file (generally has a `.spm` extension) that
+            contains the vocabulary necessary to instantiate a tokenizer.
+        sp_model_kwargs (:obj:`dict`, `optional`):
+            Will be passed to the ``SentencePieceProcessor.__init__()`` method. The `Python wrapper for SentencePiece
+            <https://github.com/google/sentencepiece/tree/master/python>`__ can be used, among other things, to set:
+
+            - ``enable_sampling``: Enable subword regularization.
+            - ``nbest_size``: Sampling parameters for unigram. Invalid for BPE-Dropout.
+
+              - ``nbest_size = {0,1}``: No sampling is performed.
+              - ``nbest_size > 1``: samples from the nbest_size results.
+              - ``nbest_size < 0``: assuming that nbest_size is infinite and samples from the all hypothesis (lattice)
+                using forward-filtering-and-backward-sampling algorithm.
+
+            - ``alpha``: Smoothing parameter for unigram sampling, and dropout probability of merge operations for
+              BPE-dropout.
+    """
+
+    def __init__(self, vocab_file: str, sp_model_kwargs: Optional[Dict[str, Any]] = None, *args, **kwargs) -> None:
+        super().__init__(sp_model_kwargs=sp_model_kwargs, *args, **kwargs)
+
+        self.vocab_file = vocab_file
+        self.sp_model_kwargs = {} if sp_model_kwargs is None else sp_model_kwargs
+        self.sp_model = self.load_spm(self.vocab_file, self.sp_model_kwargs)
+
+    def convert_tokens_to_string(self, tokens: List[str]) -> str:
+        """
+        Convert a sequence of tokens (strings for sub-words) to a single string.
+
+        This is the reverse function of ``_tokenize``.
+        """
+        return self.sp_model.decode(tokens)
+
+    def _tokenize(self, text: str) -> List[str]:
+        """
+        Encode text input to segmented tokens.
+
+        This can be revertet with the ``convert_tokens_to_string`` function.
+
+        Args:
+            text (str): the test to tokenize
+
+        Returns:
+            List[str]: the segmented tokens
+        """
+        return self.sp_model.encode(text, out_type=str)
+
+    def _tokenize_special(self, text: str) -> List[str]:
```

Comment:
Hey @LysandreJik - I would like to hear your feedback about this function.
Is it cool to refactor it into the base class? Or is it overengineered?

Thanks
Philip

Reply:
I think generally speaking we'd like to have methods that are common to *all* tokenizers in the base class - but not methods that are common to some of them only. I'd also like to keep the number of abstraction layers to a minimum, tokenizers are already quite tough to understand.